presto-main/src/main/java/com/facebook/presto/metadata/SqlScalarFunction.java

```diff
@@ -112,13 +112,29 @@ protected SqlScalarFunction(
                 literalParameters);
     }
 
+    protected SqlScalarFunction(Signature signature)
+    {
+        this.signature = requireNonNull(signature, "signature is null");
+        checkArgument(signature.getKind() == SCALAR, "function kind must be SCALAR");
+    }
+
     @Override
     public final Signature getSignature()
     {
         return signature;
     }
 
-    public abstract ScalarFunctionImplementation specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry);
+    public abstract ScalarFunctionImplementation specialize(Map<String, Type> types, List<TypeSignature> parameterTypes, TypeManager typeManager, FunctionRegistry functionRegistry);
```

Comment:
In `parameterTypes` we have only entires for types which are explicitly constrained in the `Signature` and assigned to type variable.
In case when we declare function without type variables, like below
`some_function(DECIMAL(x,y) param)` we do not have access type of `param` in `specialize` - and therefore cannot make use of values of `x` and `y`.

It seems like we could drop `parameterTypes` argument and come back to `arity` if we support using variadic binding to parametrised types (currently `variadicBound` can be `"decimal"` but cannot be `"decimal(x,y)"`).

But eventually we want to support defining function with variable number of arguments.  And infer return type based on actual arguments passed by the user - not common super type of types of all the arguments. 
We want `concat(VARCHAR(5), VARCHAR(3))` to return `VARCHAR(8)` and not `VARCHAR(10)` or `VARCHAR`.

Then `parameterTypes` list can come handy.

@sopel39 let me know if I missed something. This is originally your change.